src/main/java/org/openstreetmap/atlas/checks/validation/intersections/LineCrossingWaterBodyCheck.java

```diff
@@ -370,6 +375,41 @@ private boolean canCrossWaterBody(final AtlasItem crossingItem,
                         .allMatch(nodeHasFordTag);
     }
 
+    /**
+     * @param atlas
+     *            the atlas
+     * @param object
+     *            waterbody atlasObject
+     * @param waterbody
+     *            geometric shape of waterbody
+     * @return set of intersectingLocations
+     */
+    private Set<Location> collectIntersectingLocations(final Atlas atlas, final AtlasObject object,
```

Comment:
This seems to recompute a lot of what has already been done in the flag method, with finding intersecting line items and calling `this.getInteractionsPerWaterbodyComponent`. It also does not seem to take into account the state of `this.flagBuildings`.

Reply:
Agreed with Bentley here -- to avoid re-computation, we should be able to create a CheckFlag earlier in the flag() method and pass it to `this.getInteractionsPerWaterbodyComponent` which would store the CheckFlag with its return Set<Location>. The CheckFlag would be passed to `this.collectOffendingLineItems` as well, where it can filter out the CheckFlag of any valid intersection locations, similar to here: https://github.com/osmlab/atlas-checks/blob/b4487d840596d4eed588c176036fc176d5cddd22/src/main/java/org/openstreetmap/atlas/checks/validation/intersections/LineCrossingWaterBodyCheck.java#L440